java/vector/src/main/java/org/apache/arrow/vector/dictionary/BatchedDictionary.java

```diff
@@ -0,0 +1,270 @@
+    if (!(BaseVariableWidthVector.class.isAssignableFrom(dictionary.getClass()))) {
+      throw new IllegalArgumentException("Dictionary must be a superclass of 'BaseVariableWidthVector' " +
+          "such as 'VarCharVector'.");
+    }
+    if (dictionary.getField().isNullable()) {
+      throw new IllegalArgumentException("Dictionary must be non-nullable.");
+    }
+    this.dictionary = (BaseVariableWidthVector) dictionary;
+    if (!(BaseIntVector.class.isAssignableFrom(indexVector.getClass()))) {
+      throw new IllegalArgumentException("Index vector must be a superclass type of 'BaseIntVector' " +
+          "such as 'IntVector' or 'Uint4Vector'.");
+    }
+    this.indexVector = (BaseIntVector) indexVector;
+    hashTable = new DictionaryHashTable();
+  }
+
+  /**
+   * The index vector.
+   */
+  public FieldVector getIndexVector() {
+    return indexVector;
+  }
+
+  @Override
+  public FieldVector getVector() {
+    return dictionary;
+  }
+
+  @Override
+  public ArrowType getVectorType() {
+    return dictionary.getField().getType();
+  }
+
+  @Override
+  public DictionaryEncoding getEncoding() {
+    return encoding;
+  }
+
+  /**
+   * Considers the entire byte array as the dictionary value. If the value is null,
+   * a null will be written to the index.
+   *
+   * @param index the value to change
+   * @param value the value to write.
+   */
+  public void setSafe(int index, byte[] value) {
+    if (value == null) {
+      setNull(index);
+      return;
+    }
+    setSafe(index, value, 0, value.length);
+  }
+
+  /**
+   * Encodes the given range in the dictionary. If the value is null, a null will be
+   * written to the index.
+   *
+   * @param index the value to change
+   * @param value the value to write.
+   * @param offset An offset into the value array.
+   * @param len The length of the value to write.
+   */
+  public void setSafe(int index, byte[] value, int offset, int len) {
+    if (value == null || len == 0) {
+      setNull(index);
+      return;
+    }
+    int di = getIndex(value, offset, len);
+    indexVector.setWithPossibleTruncate(index, di);
+  }
+
+  /**
+   * Set the element at the given index to null.
+   *
+   * @param index the value to change
+   */
+  public void setNull(int index) {
+    indexVector.setNull(index);
+  }
+
+  @Override
+  public void close() throws IOException {
+    dictionary.close();
+    indexVector.close();
+  }
+
+  /**
+   * Mark the dictionary as complete for the batch. Called by the {@link ArrowWriter}
+   * on {@link ArrowWriter#writeBatch()}.
+   */
+  public void mark() {
+    dictionary.setValueCount(dictionaryIndex);
+    // not setting the index vector value count. That will happen when the user calls
+    // VectorSchemaRoot#setRowCount().
+  }
+
+  /**
+   * Resets the dictionary to be used for a new batch. Called by the {@link ArrowWriter} on
+   * {@link ArrowWriter#writeBatch()}.
+   */
+  public void reset() {
+    wasReset = true;
```

Comment:
wasReset only ever gets set to true. So if you reset() the dictionary, then populate it again, it is still marked as "wasReset". Is this intentional? It seems like the object basically becomes unusuable if it is reset.

Reply:
It's intended as a fail-fast flag so that if the user populates the dictionary, writes a batch to a file, then tries to write a replacement dictionary, the exception would be thrown before the next call to `writeBatch`. Part of the reason I left the TODO's in there.

If ya'll (and @danepitkin) think it's ok to wait until a call to `writeBatch` then I can drop that `forFileIPC` flag and catch the error in the `ArrowFileWriter`. I'd still need a flag like the reset flag somewhere to track if a dictionary was a replacement or not.

Reply:
I guess when I see a "reset" function I think it's to restore an object to an initial state so that it can be re-used.

Reply:
In this case it is used to reset the state, depending on delta (maintain memo but reset the dictionary vector so it can hold new unique values) or replacement (reset dictionary _and_ memo) _and_ stream or file mode (throw error on update before flushing a batch). Definitely confusing. I'd rather update the file spec to support replacements.